File: Gauge Set Optimization/kmeans/kmeans.py
# Required imports
import numpy as np
from numpy.random import random_integers
from sklearn.decomposition import PCA
from scipy.cluster.vq import kmeans, vq

# Selects a random subset of half the data as training data
def produce_data(data):
	shape = data.shape
	num_users = shape[0]
	num_jokes = shape[1]

	print 'Selecting training data'
	# select half the users
	random_sample = random_integers(0, high=num_users - 1, size=num_users/2)
	other = np.array([i for i in range(num_users) if i not in random_sample])
	train = data[random_sample, :] # keep the randomly chosen users
	test = data[other, :]
	print 'Done'
	return train, test

# Returns the median of a row after removing nan values
def nan_median(row):
	return np.median(row[~np.isnan(row)])

# Perform PCA and then subsequently k-means clustering
def perform_clustering(data, gauge, dim, n_clusters=20, out_of_sample=True, out_data=[]):
	print 'Performing PCA'
	cluster_data = data[:, gauge]
	pca = PCA(n_components=dim) # Create a PCA model with 4 components
	pca.fit(cluster_data) # fit the model to the data
	cluster_data = pca.transform(cluster_data) # get the new data
	print 'Done'

	print 'Performing clustering'
	code_book = kmeans(cluster_data, n_clusters, iter=30)[0]
	clusters = vq(cluster_data, code_book)[0]
	print 'Done'
	return (clusters, code_book)

# Calculate the nmae of a data and gauge set
def nmae(data, gauge, jokes=100, dim=4, n_clusters=20):
	print 'Calculating NMAE'
	# Get the cluster that each user belongs to
	clusters, code_book = perform_clustering(data, gauge, dim, n_clusters)
	shape = data.shape
	users = shape[0]
	error = 0
	ratings = 0

	for user in range(users):
		print str(user) + ' of ' + str(users)
		for joke in range(jokes): # For each joke
			if joke in gauge: # Don't rate the gauge set jokes
				continue
			cluster = clusters[user] # cluster that this user belongs to
			# Ratings from members of the same cluster
			members = data[clusters == cluster, joke]
			pred = nan_median(members) # prediction is median value
			real = data[user, joke] # real value
			diff = abs(pred - real) # absoloute value of difference
			error += diff/20.0 # add to total error
			ratings += 1

	nmae = error/ratings
	return (nmae, code_book) # Return the code book along with NMAE
	# Use code book for out of sample

def nmae_out(code_book, data, gauge, jokes=100, dim=4, n_clusters=20):
	print 'Calculating out of sample NMAE'

	print 'Performing PCA'
	cluster_data = data[:, gauge]
	pca = PCA(n_components=dim) # Create a PCA model with 4 components
	pca.fit(cluster_data) # fit the model to the data
	cluster_data = pca.transform(cluster_data) # get the new data
	print 'Done'

	clusters = vq(cluster_data, code_book)[0] # get the clusters of out of sample data
	shape = data.shape
	users = shape[0]
	error = 0
	ratings = 0

	for user in range(users):
		print str(user) + ' of ' + str(users)
		for joke in range(jokes): # For each joke
			if joke in gauge: # Don't rate the gauge set jokes
				continue
			cluster = clusters[user] # cluster that this user belongs to
			# Ratings from members of the same cluster
			members = data[clusters == cluster, joke]
			pred = nan_median(members) # prediction is median value
			real = data[user, joke] # real value
			diff = abs(pred - real) # absoloute value of difference
			error += diff/20.0 # add to total error
			ratings += 1

	nmae = error/ratings
	return (nmae, code_book) # Return the code book along with NMAE
	# Use code book for out of sample

print 'Loading data from file'
data = np.genfromtxt('../jester1.csv', delimiter=',') # Read the file
print 'Done'

print 'Defining gauge set'
gauge = np.array([13, 25, 27, 37, 38, 60]) # Predefined gauge set
# Jokes not in the gauge set
not_gauge = np.array([i for i in range(100) if i not in gauge])
print 'Done'

print 'Reshaping data set'
data = data[:, 1:] # Dropping the first column
data[data == 99] = None # Set missing values to None
print 'Done'

print 'Selecting rows to keep'
# Only keep the users that have rated all data
rows_to_keep = np.apply_along_axis(lambda row:
									reduce(lambda x, y: x and y, row), 1, ~np.isnan(data))
pca_data = data[rows_to_keep, :] # PCA_data is only the data we want to keep
print 'Done'

print 'Defining constants'
shape = pca_data.shape
num_users = shape[0]
num_jokes = shape[1]
print 'Done'

## Generating data for number of clusters vs. NMAE
def clusters_vs_nmae():
	all_nmaes = []
	all_nmaes_out = []
	clusters = []
	for i in range(15, 100, 5):
		print 'Computing NMAE for ' + str(i) + ' clusters'
		train, test = produce_data(pca_data)
		NMAE, code_book = nmae(train, gauge, n_clusters=i) # in sample nmae
		NMAE_out, code_book = nmae_out(code_book, test, gauge) # out of sample nmae
		print 'Result: ' + str(NMAE)
		clusters.append(i) # Add the number of clusters
		all_nmaes.append(NMAE) # Add the NMAE from this number of clusters
		all_nmaes_out.append(NMAE_out) # Add the out of sample NMAE
	return (clusters, all_nmaes, all_nmaes_out) # return list of all clusters and nmaes

File: Gauge Set Optimization/kmeans/kmeans.r
## Using K-means

require(compiler) # Enabling compilation
enableJIT(3)

rm(list = ls())
# Switch workspace
setwd("/Users/virajmahesh/Documents/Workspace/Eigentaste/kmeans/")
# Load the library for pam
library(cluster)

cat("Reading Data")
raw_data = read.csv("../jester1.csv", header = FALSE) # Read the jester data from the .csv file
pca_data = as.matrix(raw_data[,-1]) # Convert the data.frame to a matrix removing the first column
pca_data[pca_data == 99] <- NA # replace values of 99 with NA

gauge = c(14, 26, 28, 38, 39, 61) # Chosen gauge set
not_gauge = (1:100)[-gauge] # Jokes not in the gauge set

# Remove those users that have not rated every joke in the gauge set
for (i in 1:100) {
  pca_data = pca_data[is.finite(pca_data[, i]),]
}

# Calculate the NMAE for a given number of dimensions and clusters
NMAE<-function(n_dim, n_clusters) {
  cat("Selecting a random sample of half the users\n")
  
  users = dim(pca_data)[1] # number of users
  jokes = dim(pca_data)[2] # number of jokes
  
  selection = sample(users, users/2) # Random selection of half of the data
  train = pca_data[selection,] # Training data
  test = pca_data[-selection] # Testing data
  
  cluster_data = train[, gauge] # Data to be used for clustering
  
  cat("Performing PCA\n")
  pca = princomp(cluster_data, scores=TRUE) # Compute PCA and keep the scores
  scores = pca$scores # Get the scores after PCA
  cluster_data = scores[, 1:n_dim] # Keep n_dim dimensions
  explained_variance = sum((as.vector(pca$sdev^2/sum(pca$sdev^2)))[1:n_dim])
  cat("Performing clustering\n")
  clusters = kmeans(cluster_data, n_clusters, iter.max=20)$cluster # Perform clustering
  
  cat("Generating predictions\n")
  predictions = c() # Empty table
  
  for (i in 1:n_clusters) { # for each cluster
    members = train[clusters == i, ] # Ratings by users in the cluster for jokes not in the gauge set
    predicted_value = apply(members, 2, median, na.rm=TRUE) # Prediction for one cluster
    predictions = rbind(predictions, predicted_value) # Add prediction
  }
  
  cat("Calculation error\n")
  error = 0
  ratings = 0
  
  for (i in 1:(users/2)) {
    cat("User ", i, " of ", users/2, "\n")
    for (j in 1:jokes) {
      if (! j %in% gauge) {
        cluster = clusters[i]
        pred = predictions[cluster, j]
        real = train[i, j]
        error = error + abs(pred - real)/20.0
        ratings = ratings + 1
      }
    }
  }
  
  nmae = error/ratings
  cat("NMAE: ", nmae, "\n")
  
  c(explained_variance, nmae)
}

results = c()
for (dimension in 1:6) { # for each dimension
  for (i in 1:5) { # Take 5 observations per dimension
    results = rbind(results, NMAE(dimension, 50)) # 50 clusters, accumulate result
  }
}

plot(results, xlab='Explained Variance', ylab='NMAE')

results_2 = c()
for (dimension in 1:6) { # for each dimension
  for (i in 1:10) { # Take 5 observations per dimension
    results_2 = rbind(results_2, NMAE(dimension, 50)) # 50 clusters, accumulate result
  }
}

results_3 = c()
results_4 = c()
for (dimension in 1:6) {
  prev = (dimension-1)*10 + 1
  curr = dimension*10
  section = results_2[prev:curr,]
  avg = apply(section, 2, mean)
  med = apply(section, 2, median)
  results_3 = rbind(results_3, avg)
  results_4 = rbind(results_4, med)
}

plot(results, xlab='Explained Variance', ylab='NMAE (5 obs.)')
plot(results_2, xlab='Explained Variance', ylab='NMAE (10 obs.)')
plot(results_3, type='o', xlab='Explained Variance', ylab='NMAE (mean)')
plot(results_4, type='o', xlab='Explained Variance', ylab='NMAE (median)')